Add `#[rustc_safe_intrinsic]`

CohenArthur opened the pull request:
This PR adds the `#[rustc_safe_intrinsic]` attribute as mentionned on Zulip. The goal of this attribute is to avoid keeping a list of symbols as the source for stable intrinsics, and instead rely on an attribute. This is similar to `#[rustc_const_stable]` and `#[rustc_const_unstable]`, which among other things, are used to mark the constness of intrinsic functions.

rust-highfive commented:
r? @compiler-errors

(rust-highfive has picked a reviewer for you, use r? to override)

compiler-errors commented:
Hm, Does this need a compiler MCP?

compiler-errors commented:
r? @wesleywiser since you said you would review the patches :)

bors commented:
:umbrella: The latest upstream changes (presumably #96946) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8269013084?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a255c4b526bc908ddf9137f51d2ac8b5a10bdd90/checks/8269013084/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0133]: call to unsafe function is unsafe and requires unsafe block
   --> library/core/src/ptr/const_ptr.rs:573:20
    |
573 |         let this = intrinsics::ptr_mask(self.cast::<()>(), mask);
    |
    = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function is unsafe and requires unsafe block
error[E0133]: call to unsafe function is unsafe and requires unsafe block
   --> library/core/src/ptr/mut_ptr.rs:589:20
    |
589 |         let this = intrinsics::ptr_mask(self.cast::<()>(), mask) as *mut ();
    |
    = note: consult the function's documentation for information on how to avoid undefined behavior

For more information about this error, try `rustc --explain E0133`.
```

</details>

CohenArthur commented:
Following the discussion around the MCP, I've added the logic for checking the presence of the attribute on top of the existing "safe intrinsic list". This allows consumers of the `core` library to rely on the presence of the attribute, but keeps the compiler as the source of trust for which intrinsics are safe

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8272494408?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/422fad6e4396a2bad7558e8025f8138fa6206216/checks/8272494408/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1042 tests
---
failures:

---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0094 (line 1978) stdout ----
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Unsafe`,
 right: `Normal`: Intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `size_of`', compiler/rustc_typeck/src/check/intrinsic.rs:118:5
stack backtrace:
   0:     0x7f473194a79e - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf59a7b4e2d434620
   1:     0x7f47319b3948 - core::fmt::write::ha2091e91361753a7
   2:     0x7f473193b0a1 - std::io::Write::write_fmt::h5295384f2cf988e3
   3:     0x7f473194d7ae - std::panicking::default_hook::{{closure}}::h1a4a7c656812fb23
   4:     0x7f473194d477 - std::panicking::default_hook::h375a8262cbf87d78
   5:     0x7f4732312174 - rustc_driver[8a450591bc686d17]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7f473194df61 - std::panicking::rust_panic_with_hook::h4f679a357a9ccd79
   7:     0x7f473194dd87 - std::panicking::begin_panic_handler::{{closure}}::h4e87b160fb35c7c3
   8:     0x7f473194acd4 - std::sys_common::backtrace::__rust_end_short_backtrace::hb01360a6cd9f56db
   9:     0x7f473194da52 - rust_begin_unwind
  10:     0x7f47318fcf13 - core::panicking::panic_fmt::h8700365ed7eef2a2
  11:     0x7f47319b030e - core::panicking::assert_failed_inner::h37448dc7036e8bb4
  12:     0x7f473217594b - core[2e31ec2f599a4931]::panicking::assert_failed::<rustc_hir[a09fe2c49dca5e34]::hir::Unsafety, rustc_hir[a09fe2c49dca5e34]::hir::Unsafety>
  13:     0x7f4732e602e2 - rustc_typeck[9d52801361d57d]::check::intrinsic::intrinsic_operation_unsafety
  14:     0x7f4732d18554 - rustc_typeck[9d52801361d57d]::collect::fn_sig
  15:     0x7f4733f0130e - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::DefId, rustc_middle[ef618a27e98a180]::ty::sty::Binder<rustc_middle[ef618a27e98a180]::ty::sty::FnSig>>>
  16:     0x7f473400cbf3 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::fn_sig, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  17:     0x7f4733e0a9ac - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::fn_sig
  18:     0x7f4732d0f3e6 - rustc_typeck[9d52801361d57d]::collect::convert_item
  19:     0x7f4732d0af1a - <rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor as rustc_hir[a09fe2c49dca5e34]::intravisit::Visitor>::visit_item
  20:     0x7f4732d32180 - <rustc_middle[ef618a27e98a180]::hir::map::Map>::visit_item_likes_in_module::<rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor>
  21:     0x7f4732d0a4ad - rustc_typeck[9d52801361d57d]::collect::collect_mod_item_types
  22:     0x7f4733eedfe0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::LocalDefId, ()>>
  23:     0x7f4733fe11ef - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::collect_mod_item_types, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  24:     0x7f4733e11110 - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::collect_mod_item_types
  25:     0x7f4732d317da - <rustc_middle[ef618a27e98a180]::hir::map::Map>::for_each_module::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}::{closure#0}::{closure#0}>
  26:     0x7f4732d7da2b - <rustc_session[492dfb63ec0bb8ae]::session::Session>::track_errors::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}, ()>
  27:     0x7f4732fd6119 - rustc_typeck[9d52801361d57d]::check_crate
  28:     0x7f4732450631 - rustc_interface[53f1f933a2d2ec9]::passes::analysis
  29:     0x7f4733f2d16f - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<(), core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>>
  30:     0x7f473400d2a0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::analysis, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  31:     0x7f4733dead9a - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::analysis
  32:     0x7f47323764f6 - <rustc_interface[53f1f933a2d2ec9]::passes::QueryContext>::enter::<rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  33:     0x7f4732329751 - rustc_interface[53f1f933a2d2ec9]::interface::create_compiler_and_run::<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>
  34:     0x7f47322fd742 - <scoped_tls[537ffb9e41f61595]::ScopedKey<rustc_span[d12d2727cd9c9f44]::SessionGlobals>>::set::<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  35:     0x7f473237a6cf - std[2e4d6ca4d7177998]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  36:     0x7f473232838e - std[2e4d6ca4d7177998]::panic::catch_unwind::<core[2e31ec2f599a4931]::panic::unwind_safe::AssertUnwindSafe<<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  37:     0x7f473237db52 - <<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1} as core[2e31ec2f599a4931]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f473195adc5 - std::sys::unix::thread::Thread::new::thread_start::hbd1073ee1be91ca5
  39:     0x7f47316f6b43 - <unknown>
  40:     0x7f4731788a00 - <unknown>
  41:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-nightly (4bad79f8f 2022-09-09) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C codegen-units=1 -Z normalize-docs -Z unstable-options
query stack during panic:
query stack during panic:
#0 [fn_sig] computing function signature of `main::_doctest_main__checkout_obj_build_x86_64_unknown_linux_gnu_test_error_index_md_1978_0::size_of`
#1 [collect_mod_item_types] collecting item types in top-level module
#2 [analysis] running analysis passes on this crate
end of query stack
Some expected error codes were not found: ["E0094"]
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0094 (line 1991) stdout ----
  left: `Unsafe`,
  left: `Unsafe`,
 right: `Normal`: Intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `size_of`', compiler/rustc_typeck/src/check/intrinsic.rs:118:5
stack backtrace:
   0:     0x7fae539ce79e - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf59a7b4e2d434620
   1:     0x7fae53a37948 - core::fmt::write::ha2091e91361753a7
   2:     0x7fae539bf0a1 - std::io::Write::write_fmt::h5295384f2cf988e3
   3:     0x7fae539d17ae - std::panicking::default_hook::{{closure}}::h1a4a7c656812fb23
   4:     0x7fae539d1477 - std::panicking::default_hook::h375a8262cbf87d78
   5:     0x7fae54396174 - rustc_driver[8a450591bc686d17]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7fae539d1f61 - std::panicking::rust_panic_with_hook::h4f679a357a9ccd79
   7:     0x7fae539d1d87 - std::panicking::begin_panic_handler::{{closure}}::h4e87b160fb35c7c3
   8:     0x7fae539cecd4 - std::sys_common::backtrace::__rust_end_short_backtrace::hb01360a6cd9f56db
   9:     0x7fae539d1a52 - rust_begin_unwind
  10:     0x7fae53980f13 - core::panicking::panic_fmt::h8700365ed7eef2a2
  11:     0x7fae53a3430e - core::panicking::assert_failed_inner::h37448dc7036e8bb4
  12:     0x7fae541f994b - core[2e31ec2f599a4931]::panicking::assert_failed::<rustc_hir[a09fe2c49dca5e34]::hir::Unsafety, rustc_hir[a09fe2c49dca5e34]::hir::Unsafety>
  13:     0x7fae54ee42e2 - rustc_typeck[9d52801361d57d]::check::intrinsic::intrinsic_operation_unsafety
  14:     0x7fae54d9c554 - rustc_typeck[9d52801361d57d]::collect::fn_sig
  15:     0x7fae55f8530e - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::DefId, rustc_middle[ef618a27e98a180]::ty::sty::Binder<rustc_middle[ef618a27e98a180]::ty::sty::FnSig>>>
  16:     0x7fae56090bf3 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::fn_sig, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  17:     0x7fae55e8e9ac - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::fn_sig
  18:     0x7fae54d933e6 - rustc_typeck[9d52801361d57d]::collect::convert_item
  19:     0x7fae54d8ef1a - <rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor as rustc_hir[a09fe2c49dca5e34]::intravisit::Visitor>::visit_item
  20:     0x7fae54db6180 - <rustc_middle[ef618a27e98a180]::hir::map::Map>::visit_item_likes_in_module::<rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor>
  21:     0x7fae54d8e4ad - rustc_typeck[9d52801361d57d]::collect::collect_mod_item_types
  22:     0x7fae55f71fe0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::LocalDefId, ()>>
  23:     0x7fae560651ef - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::collect_mod_item_types, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  24:     0x7fae55e95110 - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::collect_mod_item_types
  25:     0x7fae54db57da - <rustc_middle[ef618a27e98a180]::hir::map::Map>::for_each_module::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}::{closure#0}::{closure#0}>
  26:     0x7fae54e01a2b - <rustc_session[492dfb63ec0bb8ae]::session::Session>::track_errors::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}, ()>
  27:     0x7fae5505a119 - rustc_typeck[9d52801361d57d]::check_crate
  28:     0x7fae544d4631 - rustc_interface[53f1f933a2d2ec9]::passes::analysis
  29:     0x7fae55fb116f - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<(), core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>>
  30:     0x7fae560912a0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::analysis, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  31:     0x7fae55e6ed9a - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::analysis
  32:     0x7fae543fa4f6 - <rustc_interface[53f1f933a2d2ec9]::passes::QueryContext>::enter::<rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  33:     0x7fae543ad751 - rustc_interface[53f1f933a2d2ec9]::interface::create_compiler_and_run::<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>
  34:     0x7fae54381742 - <scoped_tls[537ffb9e41f61595]::ScopedKey<rustc_span[d12d2727cd9c9f44]::SessionGlobals>>::set::<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  35:     0x7fae543fe6cf - std[2e4d6ca4d7177998]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  36:     0x7fae543ac38e - std[2e4d6ca4d7177998]::panic::catch_unwind::<core[2e31ec2f599a4931]::panic::unwind_safe::AssertUnwindSafe<<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  37:     0x7fae54401b52 - <<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1} as core[2e31ec2f599a4931]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7fae539dedc5 - std::sys::unix::thread::Thread::new::thread_start::hbd1073ee1be91ca5
  39:     0x7fae5377ab43 - <unknown>
  40:     0x7fae5380ca00 - <unknown>
  41:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-nightly (4bad79f8f 2022-09-09) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C codegen-units=1 -Z normalize-docs -Z unstable-options
query stack during panic:
query stack during panic:
#0 [fn_sig] computing function signature of `main::_doctest_main__checkout_obj_build_x86_64_unknown_linux_gnu_test_error_index_md_1991_0::size_of`
#1 [collect_mod_item_types] collecting item types in top-level module
#2 [analysis] running analysis passes on this crate
Couldn't compile the test.
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0211::_::Note__this_error_code_is_no_longer_emitted_by_the_compiler_ (line 3843) stdout ----
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Unsafe`,
  left: `Unsafe`,
 right: `Normal`: Intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `size_of`', compiler/rustc_typeck/src/check/intrinsic.rs:118:5
stack backtrace:
   0:     0x7f1d9d77579e - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf59a7b4e2d434620
   1:     0x7f1d9d7de948 - core::fmt::write::ha2091e91361753a7
   2:     0x7f1d9d7660a1 - std::io::Write::write_fmt::h5295384f2cf988e3
   3:     0x7f1d9d7787ae - std::panicking::default_hook::{{closure}}::h1a4a7c656812fb23
   4:     0x7f1d9d778477 - std::panicking::default_hook::h375a8262cbf87d78
   5:     0x7f1d9e13d174 - rustc_driver[8a450591bc686d17]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7f1d9d778f61 - std::panicking::rust_panic_with_hook::h4f679a357a9ccd79
   7:     0x7f1d9d778d87 - std::panicking::begin_panic_handler::{{closure}}::h4e87b160fb35c7c3
   8:     0x7f1d9d775cd4 - std::sys_common::backtrace::__rust_end_short_backtrace::hb01360a6cd9f56db
   9:     0x7f1d9d778a52 - rust_begin_unwind
  10:     0x7f1d9d727f13 - core::panicking::panic_fmt::h8700365ed7eef2a2
  11:     0x7f1d9d7db30e - core::panicking::assert_failed_inner::h37448dc7036e8bb4
  12:     0x7f1d9dfa094b - core[2e31ec2f599a4931]::panicking::assert_failed::<rustc_hir[a09fe2c49dca5e34]::hir::Unsafety, rustc_hir[a09fe2c49dca5e34]::hir::Unsafety>
  13:     0x7f1d9ec8b2e2 - rustc_typeck[9d52801361d57d]::check::intrinsic::intrinsic_operation_unsafety
  14:     0x7f1d9eb43554 - rustc_typeck[9d52801361d57d]::collect::fn_sig
  15:     0x7f1d9fd2c30e - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::DefId, rustc_middle[ef618a27e98a180]::ty::sty::Binder<rustc_middle[ef618a27e98a180]::ty::sty::FnSig>>>
  16:     0x7f1d9fe37bf3 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::fn_sig, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  17:     0x7f1d9fc359ac - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::fn_sig
  18:     0x7f1d9eb3a3e6 - rustc_typeck[9d52801361d57d]::collect::convert_item
  19:     0x7f1d9eb35f1a - <rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor as rustc_hir[a09fe2c49dca5e34]::intravisit::Visitor>::visit_item
  20:     0x7f1d9eb5d180 - <rustc_middle[ef618a27e98a180]::hir::map::Map>::visit_item_likes_in_module::<rustc_typeck[9d52801361d57d]::collect::CollectItemTypesVisitor>
  21:     0x7f1d9eb354ad - rustc_typeck[9d52801361d57d]::collect::collect_mod_item_types
  22:     0x7f1d9fd18fe0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<rustc_span[d12d2727cd9c9f44]::def_id::LocalDefId, ()>>
  23:     0x7f1d9fe0c1ef - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::collect_mod_item_types, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  24:     0x7f1d9fc3c110 - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::collect_mod_item_types
  25:     0x7f1d9eb5c7da - <rustc_middle[ef618a27e98a180]::hir::map::Map>::for_each_module::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}::{closure#0}::{closure#0}>
  26:     0x7f1d9eba8a2b - <rustc_session[492dfb63ec0bb8ae]::session::Session>::track_errors::<rustc_typeck[9d52801361d57d]::check_crate::{closure#0}, ()>
  27:     0x7f1d9ee01119 - rustc_typeck[9d52801361d57d]::check_crate
  28:     0x7f1d9e27b631 - rustc_interface[53f1f933a2d2ec9]::passes::analysis
  29:     0x7f1d9fd5816f - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::try_execute_query::<rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt, rustc_query_system[5bbfffd82d4c24d0]::query::caches::DefaultCache<(), core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>>
  30:     0x7f1d9fe382a0 - rustc_query_system[5bbfffd82d4c24d0]::query::plumbing::get_query::<rustc_query_impl[9ac8801b4aecdd7]::queries::analysis, rustc_query_impl[9ac8801b4aecdd7]::plumbing::QueryCtxt>
  31:     0x7f1d9fc15d9a - <rustc_query_impl[9ac8801b4aecdd7]::Queries as rustc_middle[ef618a27e98a180]::ty::query::QueryEngine>::analysis
  32:     0x7f1d9e1a14f6 - <rustc_interface[53f1f933a2d2ec9]::passes::QueryContext>::enter::<rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  33:     0x7f1d9e154751 - rustc_interface[53f1f933a2d2ec9]::interface::create_compiler_and_run::<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>
  34:     0x7f1d9e128742 - <scoped_tls[537ffb9e41f61595]::ScopedKey<rustc_span[d12d2727cd9c9f44]::SessionGlobals>>::set::<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  35:     0x7f1d9e1a56cf - std[2e4d6ca4d7177998]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  36:     0x7f1d9e15338e - std[2e4d6ca4d7177998]::panic::catch_unwind::<core[2e31ec2f599a4931]::panic::unwind_safe::AssertUnwindSafe<<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>
  37:     0x7f1d9e1a8b52 - <<std[2e4d6ca4d7177998]::thread::Builder>::spawn_unchecked_<rustc_interface[53f1f933a2d2ec9]::util::run_in_thread_pool_with_globals<rustc_interface[53f1f933a2d2ec9]::interface::run_compiler<core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>, rustc_driver[8a450591bc686d17]::run_compiler::{closure#1}>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#0}, core[2e31ec2f599a4931]::result::Result<(), rustc_errors[cf93369e4ced8f3e]::ErrorGuaranteed>>::{closure#1} as core[2e31ec2f599a4931]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f1d9d785dc5 - std::sys::unix::thread::Thread::new::thread_start::hbd1073ee1be91ca5
  39:     0x7f1d9d521b43 - <unknown>
  40:     0x7f1d9d5b3a00 - <unknown>
  41:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-nightly (4bad79f8f 2022-09-09) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C codegen-units=1 -Z normalize-docs -Z unstable-options
query stack during panic:
query stack during panic:
#0 [fn_sig] computing function signature of `main::_doctest_main__checkout_obj_build_x86_64_unknown_linux_gnu_test_error_index_md_3843_0::size_of`
#1 [collect_mod_item_types] collecting item types in top-level module
#2 [analysis] running analysis passes on this crate
Couldn't compile the test.

failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0094 (line 1978)
```

</details>

bjorn3 commented:
The first commit is a spurious duplicate of a commit that has already landed on master.

bors commented:
:umbrella: The latest upstream changes (presumably #101483) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented on compiler/rustc_typeck/src/check/intrinsic.rs:
```diff
@@ -107,14 +112,24 @@ pub fn intrinsic_operation_unsafety(intrinsic: Symbol) -> hir::Unsafety {
-    }
+    };
+
+    assert_eq!(
+        has_safe_attr,
+        is_in_list,
+        "Intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `{}`",
+        tcx.item_name(intrinsic_id)
+    );
```
I think instead of ICE-ing the compiler, it would be better to report this as an error to the user. You can do that by calling [`tcx.sess.err()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/session/struct.Session.html#method.err).

CohenArthur replied:
Sounds good - I was debating between an ICE and a regular error, and opted for an ICE as from the Zulip discussion it seems to be more of an implementation/compiler error than a user error. But a regular error makes more sense for people reimplementing the core library. Done in 062b1035698

wesleywiser commented on compiler/rustc_typeck/src/check/intrinsic.rs:
```diff
@@ -107,14 +112,25 @@ pub fn intrinsic_operation_unsafety(intrinsic: Symbol) -> hir::Unsafety {
+
+    if has_safe_attr != is_in_list {
+        tcx.sess.err(
+        DiagnosticMessage::Str(format!(
+                "intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `{}`",
+                tcx.item_name(intrinsic_id)
+    )));
```
Nit: I have no idea why rustfmt is ok with this 😄 Maybe the macro call as fn arg confuses it? I'd suggest formatting it like this unless rustfmt dislikes it:

```suggestion
            DiagnosticMessage::Str(format!(
                "intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `{}`",
                tcx.item_name(intrinsic_id)
        )));
```

wesleywiser reviewed (commented):
Would you mind also adding a UI test for the compiler error? It would be good to test both `#[rustc_safe_intrinsic]` on an intrinsic that isn't safe and not having the attribute on an intrinsic that _is_ safe. Since we call `err` and not `fatal`, I think you should be able to put both cases in the test. 

Other than that, this looks good to me! 🙂

bors commented:
:umbrella: The latest upstream changes (presumably #102051) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

CohenArthur commented:
I've fixed the formatting, added the test and fixed the merge conflicts. The error does not have any location info at the moment, so I don't mind spending a little more time on that if you think it'd be worth it @wesleywiser

bors commented:
:umbrella: The latest upstream changes (presumably #102306) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented on compiler/rustc_hir_analysis/src/check/intrinsic.rs:
```diff
@@ -106,14 +111,25 @@ pub fn intrinsic_operation_unsafety(intrinsic: Symbol) -> hir::Unsafety {
+    };
+
+    if has_safe_attr != is_in_list {
+        tcx.sess.err(
+            DiagnosticMessage::Str(format!(
+                    "intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `{}`",
+                    tcx.item_name(intrinsic_id)
+        )));
```
Ah, that's my bad. I should have pointed you toward `tcx.sess.span_err`. Sorry about that! 

I think this will generate a reasonable span for the error:

```suggestion
        tcx.sess.span_err(
            tcx.def_span(intrinsic_id),
            DiagnosticMessage::Str(format!(
                    "intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic `{}`",
                    tcx.item_name(intrinsic_id)
        )));
```

wesleywiser replied:
You're welcome! 😄

CohenArthur replied:
Done in [37bf8f8](https://github.com/rust-lang/rust/pull/100719/commits/37bf8f888cb7b8a7cb461f0d41588e500b0bb9d1). Thanks again :)

wesleywiser reviewed (approved):
It should be easy enough to get a nice span associated with the error (I believe the suggestion I left is sufficient for that to happen) so let's go ahead and do that. The rest of this looks good to me!

wesleywiser commented:
Thanks @CohenArthur!

@bors r+

bors commented:
:pushpin: Commit b1b86491f1c2bb761a06b779a3a7a936a996977a has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser b1b86491f1c2bb761a06b779a3a7a936a996977a -->
<!-- homu: {"type":"Approved","sha":"b1b86491f1c2bb761a06b779a3a7a936a996977a","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b1b86491f1c2bb761a06b779a3a7a936a996977a with merge ce7f0f1aa0f02c45cad0749e63f3086234b1f422...
<!-- homu: {"type":"BuildStarted","head_sha":"b1b86491f1c2bb761a06b779a3a7a936a996977a","merge_sha":"ce7f0f1aa0f02c45cad0749e63f3086234b1f422"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3146194460/jobs/5116507118)
Approved by: wesleywiser
Pushing ce7f0f1aa0f02c45cad0749e63f3086234b1f422 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3146194460/jobs/5116507118"},"merge_sha":"ce7f0f1aa0f02c45cad0749e63f3086234b1f422"} -->

rust-timer commented:
Finished benchmarking commit (ce7f0f1aa0f02c45cad0749e63f3086234b1f422): [comparison URL](https://perf.rust-lang.org/compare.html?start=307dd938d7937c93dc7a880e81a2383b82f30ad0&end=ce7f0f1aa0f02c45cad0749e63f3086234b1f422&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=307dd938d7937c93dc7a880e81a2383b82f30ad0&end=ce7f0f1aa0f02c45cad0749e63f3086234b1f422&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 1.4%     | [1.2%, 1.6%] | 6         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -        | -     | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=307dd938d7937c93dc7a880e81a2383b82f30ad0&end=ce7f0f1aa0f02c45cad0749e63f3086234b1f422&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 3.9%     | [3.9%, 3.9%] | 1         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=307dd938d7937c93dc7a880e81a2383b82f30ad0&end=ce7f0f1aa0f02c45cad0749e63f3086234b1f422&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*
